source/server/win/server_setup_winhttp.c

```diff
@@ -0,0 +1,186 @@
+/*!
+ * @file server_setup_winhttp.c
+ * @brief Contains functionality that allows for dispatching of HTTP(s) commands via WinHTTP
+ * @remark This file doesn't use precompiled headers because metsrv.h includes a bunch of
+ *         of definitions that clash with those found in winhttp.h. Hooray Win32 API. I hate you.
+ */
+#include "../../common/common.h"
+#include <winhttp.h>
+
+#ifdef USE_WINHTTP
+
+#define HOSTNAME_LEN 512
+#define URLPATH_LEN 1024
+
+DWORD server_dispatch_http_winhttp(Remote* remote, THREAD* serverThread, int iExpirationTimeout, int iCommTimeout,
+	wchar_t* pMetUA, wchar_t* pMetProxy, wchar_t* pMetProxyUser, wchar_t* pMetProxyPass)
+{
+	BOOL running = TRUE;
+	LONG result = ERROR_SUCCESS;
+	Packet * packet = NULL;
+	THREAD * cpt = NULL;
+	URL_COMPONENTS bits;
+	DWORD ecount = 0;
+	DWORD delay = 0;
+	wchar_t tmpHostName[512];
+	wchar_t tmpUrlPath[1024];
+
+	remote->expiration_time = 0;
+	if (iExpirationTimeout > 0)
+	{
+		remote->expiration_time = current_unix_timestamp() + iExpirationTimeout;
+	}
+
+	remote->comm_timeout = iCommTimeout;
+	remote->start_time = current_unix_timestamp();
+	remote->comm_last_packet = current_unix_timestamp();
+
+	// Allocate the top-level handle
+	if (!wcscmp(pMetProxy, L"METERPRETER_PROXY"))
+	{
+		remote->hInternet = WinHttpOpen(pMetUA, WINHTTP_ACCESS_TYPE_DEFAULT_PROXY, NULL, NULL, 0);
+	}
+	else
+	{
+		remote->hInternet = WinHttpOpen(pMetUA, WINHTTP_ACCESS_TYPE_NAMED_PROXY, pMetProxy, NULL, 0);
+	}
+
+	if (!remote->hInternet)
+	{
+		dprintf("[DISPATCH] Failed WinHttpOpen: %d", GetLastError());
+		return 0;
+	}
+
+	// Proxy auth, if required.
+	if (!(wcscmp(pMetProxyUser, L"METERPRETER_USERNAME_PROXY") == 0))
+	{
+		if (!WinHttpSetOption(remote->hInternet, WINHTTP_OPTION_PROXY_USERNAME, pMetProxyUser, lstrlen(pMetProxyUser)))
+		{
+			dprintf("[DISPATCH] Failed to set proxy username");
+		}
+		if (!WinHttpSetOption(remote->hInternet, WINHTTP_OPTION_PROXY_PASSWORD, pMetProxyPass, lstrlen(pMetProxyPass)))
```

Comment:
This should probably be two if blocks, one for user, and one for password (they can be set separately).